File: Doit_025.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

/**
 * dfs 탐색을 하면서 depth 가 5인 경우가 있으면 1을 출력하면 된다.
 */
public class Doit_025 {

    static List<Integer>[] graph;
    static boolean[] visited;
    static int node;
    static int edge;
    static boolean flag;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        node = Integer.parseInt(st.nextToken());
        edge = Integer.parseInt(st.nextToken());

        graph = new ArrayList[node];
        visited = new boolean[node];

        for (int idx = 0; idx < node; idx++) {
            graph[idx] = new ArrayList<>();
        }

        for (int idx = 1; idx <= edge; idx++) {
            st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());

            graph[start].add(end);
            graph[end].add(start);
        }


        for (int idx = 0; idx < node; idx++) {
            dfs(idx, 1);
            if (flag)
                break;
        }
        if (flag) {
            System.out.println(1);
        } else {
            System.out.println(0);
        }
    }

    static void dfs(int start, int depth) {
        if (depth == 5 || flag) {
            flag = true;
            return;
        }
        if (visited[start]) {
            return;
        }
        visited[start] = true;
        for (int next : graph[start]) {
            if (!visited[next]) {
                dfs(next, depth + 1);
            }
        }
        visited[start] = false;
    }

}

File: Doit_025.cpp
#include <iostream>
#include <vector>
using namespace std;

static vector<vector<int> > vec;
static vector<bool> visited;
static bool check;

void DFS(int num, int depth) {
  if (depth == 5 || check) {
    check = true;  // check 변수를 true로 설정
    return;
  }
  visited[num] = true;
  for (int i : vec[num]) {
    if (!visited[i]) DFS(i, depth + 1);
  }
  visited[num] = false;
}

int main() {
  int n, m;
  cin >> n >> m;
  check = false;
  vec.resize(n + 1);  // vec 크기 조절, 노드 번호 범위를 고려하여 n + 1로 설정
  visited = vector<bool>(n + 1, false);

  for (int i = 0; i < m; i++) {
    int start, end;

    cin >> start >> end;
    vec[start].push_back(end);
    vec[end].push_back(start);
  }

  for (int i = 1; i <= n; i++) {  // 모든 노드에서 DFS를 시작
    DFS(i, 1);
    if (check) break;
  }

  if (check) {
    cout << "1"
         << "\n";
  } else {
    cout << "0"
         << "\n";
  }

  return 0;
}
